File: sdk/ios/Sources/OpenIncidentRUM/OpenIncidentRUM.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif

/// Open Incident — real user monitoring, in an iOS application.
///
/// The same two rules as the browser SDK, because they are the rules of any
/// library that ships inside somebody else's product:
///
///  1. **It never throws into the host application.** There is no `try` in the
///     public surface and no force-unwrap anywhere below. A failure here costs
///     its own measurement and nothing else.
///  2. **It never blocks.** Every public call hands work to a serial queue and
///     returns; nothing is encoded, and no request is made, on the thread that
///     is drawing the screen.
///
/// What it does **not** do, said here rather than discovered later: it does not
/// capture crashes. Doing that properly means signal handlers and a Mach
/// exception port, a written-to-disk report replayed on the next launch, and
/// symbolication — a subsystem, not a feature. An SDK that claims crash
/// reporting and delivers a `try/catch` is worse than one that says it has
/// none, because somebody stops looking for a real one. Report what you catch
/// with `error(_:)`.
public enum OpenIncidentRUM {

    // MARK: - Starting

    /// Starts reporting. Safe to call twice; the second call is ignored.
    ///
    /// - Parameters:
    ///   - applicationId: The RUM application id, from **Telemetry → Connect**.
    ///     It is a public write-only token — it ships inside your binary and
    ///     anybody can read it out. The workspace has to allow mobile
    ///     applications before it is accepted at all.
    ///   - endpoint: The ingestion address, without a path.
    ///   - sampleRate: Share of sessions reported, drawn once per session.
    public static func start(
        applicationId: String,
        endpoint: URL,
        sampleRate: Double = 1,
        service: String = ""
    ) {
        queue.async {
            guard client == nil, !applicationId.isEmpty else { return }
            let session = Session.resume(sampleRate: sampleRate)
            guard session.sampled else { return }
            client = Client(applicationId: applicationId, endpoint: endpoint, session: session, service: service)
            observeLifecycle()
            reportLaunch()
        }
    }

    // MARK: - What the application reports

    /// A screen the user is now looking at.
    ///
    /// Also closes the previous one: the time between two calls is what
    /// `screen_load` measures, which is the number a user feels after a tap.
    public static func screen(_ name: String) {
        let at = Date()
        queue.async {
            guard let c = client else { return }
            Session.touch()
            if let opened = screenOpenedAt, !currentScreen.isEmpty {
                let ms = at.timeIntervalSince(opened) * 1000
                // Only the first paint of a screen is a load. A screen the user
                // sat on for four minutes is not a four-minute load, and the
                // previous version of this reported exactly that.
                if !measured.contains(currentScreen) {
                    measured.insert(currentScreen)
                    c.add(vital("screen_load", ms, screen: currentScreen, at: at))
                }
            }
            currentScreen = name
            screenOpenedAt = at
            c.newView()
            c.add(Event(kind: .screenView, at: at, screen: name))
        }
    }

    /// Anything the application wants on the timeline: a tap, a step, a purchase.
    public static func action(_ name: String, attributes: [String: String] = [:]) {
        let at = Date()
        queue.async {
            guard let c = client else { return }
            Session.touch()
            c.add(Event(kind: .action, at: at, screen: currentScreen, message: name, attributes: attributes))
        }
    }

    /// Something that went wrong and was caught.
    public static func error(_ error: Error, type: String = "", stack: String = "") {
        report(type: type.isEmpty ? String(describing: Swift.type(of: error)) : type,
               message: error.localizedDescription,
               stack: stack)
    }

    /// The same, when there is no `Error` to hand.
    public static func error(type: String, message: String, stack: String = "") {
        report(type: type, message: message, stack: stack)
    }

    /// One network call, with what came back.
    ///
    /// Reported by the application rather than intercepted. Interception means
    /// a `URLProtocol` that re-issues every request through a session of ours,
    /// which changes the behaviour of uploads, streaming and background
    /// transfers in an application nobody here can test. The trade is honest:
    /// a line of code per call site, and no chance of breaking the network of
    /// the app we were added to measure.
    public static func resource(
        url: String,
        method: String = "GET",
        status: Int = 0,
        durationMs: Double = 0,
        error: String = ""
    ) {
        let at = Date()
        queue.async {
            guard let c = client else { return }
            Session.touch()
            var attributes = ["http.method": method, "http.url": url]
            if status > 0 { attributes["http.status_code"] = String(status) }
            if !error.isEmpty { attributes["error"] = error }
            c.add(Event(
                kind: .resource, at: at, screen: currentScreen,
                value: durationMs, message: url, attributes: attributes
            ))
        }
    }

    /// Name the person, however the application knows them. Hashed on arrival,
    /// with the workspace's own salt, and the value is thrown away.
    public static func identify(_ value: String?) {
        queue.async { user = value }
    }

    /// Sends whatever is buffered. Called for you when the app goes to the
    /// background; exposed because a test wants to be sure.
    public static func flush() {
        queue.async { client?.flush() }
    }

    // MARK: - Inside

    private static let queue = DispatchQueue(label: "com.openincident.rum", qos: .utility)
    private static var client: Client?
    private static var user: String?
    private static var currentScreen = ""
    private static var screenOpenedAt: Date?
    private static var measured = Set<String>()
    private static var launched = false

    static func userValue() -> String? { user }

    private static func report(type: String, message: String, stack: String) {
        let at = Date()
        queue.async {
            guard let c = client else { return }
            Session.touch()
            c.add(Event(kind: .error, at: at, screen: currentScreen,
                        errorType: type, message: message, stack: stack))
        }
    }

    private static func vital(_ name: String, _ ms: Double, screen: String, at: Date) -> Event {
        Event(kind: .vital, at: at, screen: screen, vital: name,
              value: ms.rounded(), rating: Rating.of(name, ms))
    }

    /// Cold start: from the moment the process began to the moment the first
    /// frame is on the glass.
    ///
    /// `kinfo_proc` rather than a timestamp taken in `start()`, because by then
    /// the expensive part — dyld, the runtime, the app delegate — has already
    /// happened, and a launch metric that starts after the launch measures
    /// nothing. The first frame is taken from the next main-queue turn after
    /// the window is up, which is the closest a library can get without
    /// swizzling the view controller.
    private static func reportLaunch() {
        guard !launched else { return }
        launched = true
        guard let began = processStart() else { return }
        DispatchQueue.main.async {
            let ms = Date().timeIntervalSince(began) * 1000
            // A process resumed from a saved state can report hours. Anything
            // beyond a minute is not a launch, it is a clock.
            guard ms > 0, ms < 60_000 else { return }
            queue.async {
                client?.add(vital("app_start", ms, screen: currentScreen, at: Date()))
            }
        }
    }

    private static func processStart() -> Date? {
        var info = kinfo_proc()
        var size = MemoryLayout<kinfo_proc>.stride
        var name: [Int32] = [CTL_KERN, KERN_PROC, KERN_PROC_PID, getpid()]
        let ok = sysctl(&name, u_int(name.count), &info, &size, nil, 0) == 0
        guard ok else { return nil }
        let started = info.kp_proc.p_starttime
        return Date(timeIntervalSince1970: Double(started.tv_sec) + Double(started.tv_usec) / 1_000_000)
    }

    private static func observeLifecycle() {
        #if canImport(UIKit)
        let centre = NotificationCenter.default
        // Flushed when the app leaves the foreground, which on a phone is the
        // equivalent of the browser's `pagehide`: it is the last moment the
        // process is reliably allowed to make a request.
        centre.addObserver(forName: UIApplication.didEnterBackgroundNotification, object: nil, queue: nil) { _ in
            flush()
        }
        centre.addObserver(forName: UIApplication.willTerminateNotification, object: nil, queue: nil) { _ in
            flush()
        }
        centre.addObserver(forName: UIApplication.didBecomeActiveNotification, object: nil, queue: nil) { _ in
            queue.async { Session.touch() }
        }
        #endif
    }
}

File: sdk/android/rum/src/main/java/dev/openincident/rum/OpenIncidentRum.kt
package dev.openincident.rum

import android.app.Activity
import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.Process
import android.os.SystemClock

/**
 * Open Incident — real user monitoring, in an Android application.
 *
 * The same two rules as the iOS and browser SDKs, because they are the rules of
 * any library that ships inside somebody else's product:
 *
 *  1. **It never throws into the host application.** Every public entry point
 *     swallows its own failure; a mistake here costs its measurement and
 *     nothing else.
 *  2. **It never blocks.** Every call hands work to one background thread and
 *     returns. Nothing is encoded, and no request is made, on the thread that
 *     is drawing the screen.
 *
 * What it does **not** do, said here rather than discovered later: it does not
 * capture crashes. An `UncaughtExceptionHandler` would catch a Java throw and
 * miss every native one, would have to write to disk and replay on the next
 * launch to be trustworthy, and would fight whatever crash reporter the
 * application already installs. An SDK that claims crash reporting and
 * delivers a try/catch is worse than one that says it has none, because
 * somebody stops looking for a real one. Report what you catch with [error].
 */
object OpenIncidentRum {

    // MARK: starting

    /**
     * Starts reporting. Safe to call twice; the second call is ignored.
     *
     * [applicationId] is a public write-only token: it ships inside your APK
     * and anybody can read it out. The workspace has to allow mobile
     * applications before it is accepted at all.
     */
    @JvmStatic
    @JvmOverloads
    fun start(
        context: Context,
        applicationId: String,
        endpoint: String,
        sampleRate: Double = 1.0,
        service: String = "",
    ) {
        try {
            if (client != null || applicationId.isEmpty()) return
            val app = context.applicationContext
            store = app.getSharedPreferences("oi.rum", Context.MODE_PRIVATE)
            val session = Session.resume(store!!, sampleRate)
            if (!session.sampled) return
            client = Client(app, applicationId, endpoint, session, service)
            (app as? Application)?.registerActivityLifecycleCallbacks(Lifecycle)
            reportLaunch()
        } catch (t: Throwable) {
            // A monitoring library that takes the application down with it has
            // done more harm than the absence of monitoring ever could.
        }
    }

    // MARK: what the application reports

    /**
     * A screen the user is now looking at.
     *
     * Also closes the previous one: the time between two calls is what
     * `screen_load` measures, which is the number a user feels after a tap.
     */
    @JvmStatic
    fun screen(name: String) {
        val at = System.currentTimeMillis()
        submit {
            val c = client ?: return@submit
            store?.let { Session.touch(it, at) }
            val opened = screenOpenedAt
            if (opened != null && currentScreen.isNotEmpty() && currentScreen !in measured) {
                // Only the first paint of a screen is a load. A screen the user
                // sat on for four minutes is not a four-minute load.
                measured.add(currentScreen)
                c.add(vital("screen_load", (at - opened).toDouble(), currentScreen, at))
            }
            currentScreen = name
            screenOpenedAt = at
            c.newView()
            c.add(Event(Event.Kind.SCREEN_VIEW, at, screen = name))
        }
    }

    /** Anything the application wants on the timeline: a tap, a step, a purchase. */
    @JvmStatic
    @JvmOverloads
    fun action(name: String, attributes: Map<String, String> = emptyMap()) {
        val at = System.currentTimeMillis()
        submit {
            val c = client ?: return@submit
            store?.let { Session.touch(it, at) }
            c.add(Event(Event.Kind.ACTION, at, screen = currentScreen, message = name, attributes = attributes))
        }
    }

    /** Something that went wrong and was caught. */
    @JvmStatic
    @JvmOverloads
    fun error(throwable: Throwable, type: String = "") {
        report(
            type = type.ifEmpty { throwable.javaClass.simpleName },
            message = throwable.message ?: throwable.toString(),
            stack = throwable.stackTrace.take(40).joinToString("\n") { it.toString() },
        )
    }

    /** The same, when there is no throwable to hand. */
    @JvmStatic
    @JvmOverloads
    fun error(type: String, message: String, stack: String = "") = report(type, message, stack)

    /**
     * One network call, with what came back.
     *
     * Reported by the application rather than intercepted. Interception means
     * an OkHttp interceptor we install into a client we do not own, or a
     * `URL.setURLStreamHandlerFactory` that can be set once per process and
     * will fight whatever else wanted it. The trade is honest: a line of code
     * per call site, and no chance of breaking the network of the app we were
     * added to measure.
     */
    @JvmStatic
    @JvmOverloads
    fun resource(
        url: String,
        method: String = "GET",
        status: Int = 0,
        durationMs: Double = 0.0,
        error: String = "",
    ) {
        val at = System.currentTimeMillis()
        submit {
            val c = client ?: return@submit
            store?.let { Session.touch(it, at) }
            val attributes = mutableMapOf("http.method" to method, "http.url" to url)
            if (status > 0) attributes["http.status_code"] = status.toString()
            if (error.isNotEmpty()) attributes["error"] = error
            c.add(Event(Event.Kind.RESOURCE, at, screen = currentScreen,
                value = durationMs, message = url, attributes = attributes))
        }
    }

    /**
     * Name the person, however the application knows them. Hashed on arrival,
     * with the workspace's own salt, and the value is thrown away.
     */
    @JvmStatic
    fun identify(value: String?) = submit { user = value }

    /**
     * Sends whatever is buffered. Called for you when the application stops;
     * exposed because a test wants to be sure.
     */
    @JvmStatic
    fun flush() = submit { client?.flush() }

    // MARK: inside

    private var client: Client? = null
    private var store: SharedPreferences? = null
    private var user: String? = null
    private var currentScreen = ""
    private var screenOpenedAt: Long? = null
    private val measured = mutableSetOf<String>()
    private var launched = false

    internal fun userValue(): String? = user

    private fun submit(work: () -> Unit) {
        try {
            Worker.run {
                try {
                    work()
                } catch (t: Throwable) {
                    // As above: never into the host application.
                }
            }
        } catch (t: Throwable) {
        }
    }

    private fun report(type: String, message: String, stack: String) {
        val at = System.currentTimeMillis()
        submit {
            val c = client ?: return@submit
            store?.let { Session.touch(it, at) }
            c.add(Event(Event.Kind.ERROR, at, screen = currentScreen,
                errorType = type, message = message, stack = stack))
        }
    }

    private fun vital(name: String, ms: Double, screen: String, at: Long) =
        Event(Event.Kind.VITAL, at, screen = screen, vital = name,
            value = Math.round(ms).toDouble(), rating = Rating.of(name, ms))

    /**
     * Cold start: from the moment the process began to the moment the first
     * frame is on the glass.
     *
     * `Process.getStartUptimeMillis` rather than a timestamp taken in [start],
     * because by then the expensive part — the zygote fork, the class loading,
     * the Application object — has already happened, and a launch metric that
     * starts after the launch measures nothing. The first frame is taken from
     * the next main-looper turn, which is the closest a library can get without
     * attaching to the choreographer of an activity it does not own.
     */
    private fun reportLaunch() {
        if (launched) return
        launched = true
        Handler(Looper.getMainLooper()).post {
            val ms = try {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    (SystemClock.uptimeMillis() - Process.getStartUptimeMillis()).toDouble()
                } else {
                    return@post
                }
            } catch (t: Throwable) {
                return@post
            }
            // A process the system kept warm can report hours. Beyond a minute
            // it is not a launch, it is a clock.
            if (ms <= 0 || ms >= 60_000) return@post
            submit { client?.add(vital("app_start", ms, currentScreen, System.currentTimeMillis())) }
        }
    }

    /**
     * Screens named for free, and the flush when the application leaves.
     *
     * Registered only if the context given to [start] is the `Application`, so
     * an SDK started from a library context does not silently do nothing
     * different — it does nothing extra, and [screen] still works by hand.
     */
    private object Lifecycle : Application.ActivityLifecycleCallbacks {
        private var visible = 0
        override fun onActivityCreated(a: Activity, b: Bundle?) {}
        override fun onActivityStarted(a: Activity) {
            visible++
        }
        override fun onActivityResumed(a: Activity) {
            store?.let { Session.touch(it) }
        }
        override fun onActivityPaused(a: Activity) {}
        override fun onActivityStopped(a: Activity) {
            visible--
            // Zero visible activities is the moment Android stops promising the
            // process anything, which makes it this platform's `pagehide`.
            if (visible <= 0) flush()
        }
        override fun onActivitySaveInstanceState(a: Activity, b: Bundle) {}
        override fun onActivityDestroyed(a: Activity) {}
    }
}
